client/commands/v2/persistent.py

```diff
@@ -397,6 +397,168 @@ async def run(self) -> int:
             await self.pyre_manager.ensure_task_stop()
 
 
+class PysaServer:
+    # I/O Channels
+    input_channel: connection.TextReader
+    output_channel: connection.TextWriter
+
+    # Immutable States
+    client_capabilities: lsp.ClientCapabilities
+
+    # Mutable States
+    state: ServerState
+    pyre_manager: connection.BackgroundTaskManager
+
+    def __init__(
+        self,
+        input_channel: connection.TextReader,
+        output_channel: connection.TextWriter,
+        client_capabilities: lsp.ClientCapabilities,
+        state: ServerState,
+        pyre_manager: connection.BackgroundTaskManager,
+    ) -> None:
+        self.input_channel = input_channel
+        self.output_channel = output_channel
+        self.client_capabilities = client_capabilities
+        self.state = state
+        self.pyre_manager = pyre_manager
+
+    async def wait_for_exit(self) -> int:
+        while True:
+            async with _read_lsp_request(
+                self.input_channel, self.output_channel
+            ) as request:
+                LOG.debug(f"Received post-shutdown request: {request}")
+
+                if request.method == "exit":
+                    return 0
+                else:
+                    raise json_rpc.InvalidRequestError("LSP server has been shut down")
+
+    async def _try_restart_pyre_server(self) -> None:
+        if self.state.consecutive_start_failure < CONSECUTIVE_START_ATTEMPT_THRESHOLD:
+            await self.pyre_manager.ensure_task_running()
+        else:
+            LOG.info(
+                "Not restarting Pysa since failed consecutive start attempt limit"
```

Comment:
FYI, we just added the FixIt linter internally, which I don't think is running on the open source CI. It doesn't like implicit string concatenation. This will need to be fixed here and in a few other places (I'll comment for now, but you should try to get FixIt up and running in your environment. It's this rule specifically that's biting you: https://fixit.readthedocs.io/en/latest/rules/UsePlusForStringConcatRule.html
```suggestion
                ("Not restarting Pysa since failed consecutive start attempt limit"
                + " has been reached.")
```